### ISSUE TITLE
DELETE registry blob returns BLOB_UNKNOWN

### ISSUE THREAD

**Issue by renflo** [docker/distribution#1505]
Hi,
We´ve been running a docker registry (registry:2) for a while and it´s time we clean up a bit. Some images previously used are not anymore and I want to get rid of them.

I wrote a shell script which based on a given image name and tag queries the manifest to get its manifest_digest and a list of layer_digests. Then the program loops through all layers, deleting them and when it´s done the manifest is itself deleted.

I´ve been enabling delete in the configuration (passed as environment variable on the docker container), and i´t set up correctly because the manifests get deleted.

But I cannot manage to delete the blobs.

Here is a sample blob deletion request:
```bash
curl --fail -s -X DELETE https://myhost.com:5000/v2/myapp/blobs/sha256:03e1855d4f316edea9545408dcac38be93e9ea6aba6e85610edf76db7ccbbfa7
```
I have been checking on disk (which is where we store blobs) and I find a directory which seems to "be" this blob 
```
ls /.../docker-registry/docker/registry/v2/blobs/sha256/03/03e1855d4f316edea9545408dcac38be93e9ea6aba6e85610edf76db7ccbbfa7/data
/.../docker-registry/docker/registry/v2/blobs/sha256/03/03e1855d4f316edea9545408dcac38be93e9ea6aba6e85610edf76db7ccbbfa7/data
```

Any clues what´s wrong?

Here is the verbose output from curl (which our script uses)
```
 curl -v -X DELETE https://myhost.com:5000/v2/myapp/blobs/sha256:03e1855d4f316edea9545408dcac38be93e9ea6aba6e85610edf76db7ccbbfa7
(...SSL Stuff....)
> DELETE /v2/myapp/blobs/sha256:03e1855d4f316edea9545408dcac38be93e9ea6aba6e85610edf76db7ccbbfa7 HTTP/1.1
> User-Agent: curl/7.38.0
> Host: myhost.com:5000
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< X-Content-Type-Options: nosniff
< Date: Thu, 03 Mar 2016 13:05:31 GMT
< Content-Length: 74
<
{"errors":[{"code":"BLOB_UNKNOWN","message":"blob unknown to registry"}]}
```

**Comment by dmp42** [comment]
Have you looked at the garbage collector? https://github.com/docker/distribution/pull/1386

**Comment by RichardScothern** [comment]
hi @renflo   The delete will remove the link file (soft delete) which is actually in the repository.  I'd need to see that directory to see the problem.

Your approach may corrupt data as blobs can be referenced from multiple repositories.

You can delete your manifests by digest and then run the garbage collector (see https://github.com/docker/distribution/blob/master/docs/gc.md)

**Comment by renflo** [comment]
Thanks for the pointer to the garbage collector @dmp42, I did not see this one. I´m going to read the thread, it´s quite long and quite new so I´ll keep following it until it stabilizes. It shows up in 2.4 and so I´ll wait for it to be published on docker hub. I have not been running the registry from images I created myself. How safe would it be to check out 2.4, build it and use it now instead of our current 2.3?

I was aware that the blobs could be referenced from within the same repository but not that they could also be global to the registry @RichardScothern. I knew my approach could corrupt other tags within the same repository, but it was acceptable as I was anyways going to erase the whole repository. Thanks for your comment, it will probably spare me from some problems later.

Deleting the manifest and running the garbage collector sounds great, and much, much safer. Looking forward to it.

Any prognostics on when registry:2.4 will reach docker hub? weeks? months?

**Comment by dmp42** [comment]
@renflo registry 2.4 will be released in about 5 weeks.
When the hub will catch-up on that is decided at this point, but it's definitely more than 5w from now.

**Comment by renflo** [comment]
Before I close this issue:  @RichardScothern you wrote that you´d have to see the directory to diagnose the problem. What would have you looked for? It sounds like you had some hypothesis on what the error could have been. 

I´m all ears if you want to share them with me, if it´s something I did which can have corrupted the registry I would like to learn what.

I have by the way never been modifying anything manually in the file storage. Only used API v2

**Comment by renflo** [comment]
thx @dmp42 I´ll keep an eye on it.

**Comment by RichardScothern** [comment]
@renflo 
If you post your registry logs for the delete operation, along with the output of
`find /.../docker-registry/docker/registry/v2/repositories/myapp -ls`, we'll get an idea of what is happening

You can pull and run `distribution/registry:latest` It should have the garbage collector command.

**Comment by renflo** [comment]
I´ll test `distribution/registry:latest`, thx @RichardScothern.

The clean up we´re doing is prior to the first release of the application containers we are working on. Our registry was using much more space than we think it should (tens of gigs) so we wanted to shrink it before valuable artifacts got pushed there (the kind of experimental deleting I described will not be acceptable later on).

So the drastic decision we took was to recreate a new file storage, which we did earlier today. It was both good enough and very simple. The old stuff has been deleted so I think I´ll have to accept not knowing what caused the errors.
